forem/forem
# Add Polymorphism to Broadcasts (Part 2)

## What type of PR is this? (check all applicable)


## Description
This PR is part two of a three-part refactor that creates a polymorphic association between Broadcasts, Announcements, and Welcome Notifications. This PR removes unused columns from the `Broadcasts` table, adds a `data_update` script to properly backfill Broadcast-related data. Additionally, this PR removes `WelcomeNotification` and `Announcement`-specific logic from the `Broadcasts` model to the `WelcomeNotification` and `Announcement` models, with the necessary validations. It also adjusts and adds specs around these model changes. Furthermore, this PR refactors the `Broadcasts::Controller` to account for the polymorphic association and adjusts the existing `broadcasts_controller` spec to account for this. This PR also adds the additional factories, `WelcomeNotification` and `Announcement` factories, needed to set up the polymorphic association in specs.

## Related Tickets & Documents
Related to [Add Polymorphism to Broadcasts (Part 1)](https://github.com/forem/forem/pull/9231)
Related to [Remove banner_style from Broadcasts (Part 3 of Polymorphic Broadcasts)](https://github.com/forem/forem/pull/9545)
Reason for changes are due to #7169 and #8861 

## QA Instructions, Screenshots, Recordings
To QA these changes, navigate to `/internal/broadcasts` and ensure that you can create and update Broadcasts without any issues. Additionally, ensure that you are still able to receive Welcome Notifications as a newly created (less than 7 days old) user.

**To test that you are still able to receive Welcome Notifications, you can do the following:**
- Update your user's `created_at` timestamp to be less than or equal to 7 days old:
```
me = User.last
me.created_at = 3.days.ago
me.save!
```
- Create a valid `WelcomeNotification` and `Broadcast` OR use one of the seed Welcome Notifications. For this example, I created my own `WelcomeNotification` and `Broadcast`:
```
[6] pry(main)> wn = WelcomeNotification.create
   (0.2ms)  BEGIN
  WelcomeNotification Create (1.0ms)  INSERT INTO "welcome_notifications" ("created_at", "updated_at") VALUES ($1, $2) RETURNING "id"  [["created_at", "2020-07-27 15:06:15.140527"], ["updated_at", "2020-07-27 15:06:15.140527"]]
   (0.4ms)  COMMIT
=> #<WelcomeNotification:0x00007f8f86a0d6e8
 id: 1,
 created_at: Mon, 27 Jul 2020 15:06:15 UTC +00:00,
 updated_at: Mon, 27 Jul 2020 15:06:15 UTC +00:00>
```
```
[22] pry(main)> broadcast
=> #<Broadcast:0x00007f8f8c60e258
 id: 1,
 active: true,
 active_status_updated_at: Mon, 27 Jul 2020 15:09:07 UTC +00:00,
 banner_style: nil,
 body_markdown: "Hello, test notification!",
 broadcastable_id: 1,
 broadcastable_type: "WelcomeNotification",
 created_at: Mon, 27 Jul 2020 15:09:07 UTC +00:00,
 processed_html: "Hello, test notification!",
 title: "Test Notification",
 updated_at: Mon, 27 Jul 2020 15:09:07 UTC +00:00>
```
- Send the `WelcomeNotification` (_heads up: you will need Sidekiq running to properly do this_):
```
[25] pry(main)> Notification.send_welcome_notification(me.id, broadcast.id)
=> "81d03f26c55c3f07fb0c5a95"
```
- Check to see that you received the expected Welcome Notification in the console:
```
[29] pry(main)> me.notifications.all
  Notification Load (0.7ms)  SELECT "notifications".* FROM "notifications" WHERE "notifications"."user_id" = $1  [["user_id", 1]]
=> [#<Notification:0x00007f8f8ed22358
  id: 1,
  action: "WelcomeNotification",
  created_at: Mon, 27 Jul 2020 15:10:11 UTC +00:00,
  json_data:
   {"user"=>
     {"id"=>1,
      "name"=>"Firstname Lastname",
      "path"=>"/username",
      "class"=>{"name"=>"User"},
      "username"=>"username",
      "created_at"=>"2020-07-24T15:09:47.214Z",
      "comments_count"=>0,
      "profile_image_90"=>"/uploads/user/profile_image/1/0d9357a3-6016-4b07-8941-b9598ccdd0e0.jpg"},
    "broadcast"=>
     {"title"=>"Test Notification", "processed_html"=>"Hello, test notification!", "broadcastable_type"=>"WelcomeNotification"}},
  notifiable_id: 1,
  notifiable_type: "Broadcast",
  notified_at: Mon, 27 Jul 2020 15:10:11 UTC +00:00,
  organization_id: nil,
  read: false,
  updated_at: Mon, 27 Jul 2020 15:10:11 UTC +00:00,
  user_id: 1>]
```
**To check whether or not you received the expected Welcome Notification locally, navigate to `/notifications` and double-check that the Welcome Notification is there:**
<img width="679" alt="Screen Shot 2020-07-27 at 9 23 37 AM" src="https://user-images.githubusercontent.com/32834804/88560150-ffda1280-cfea-11ea-8c16-e15b8b39d0a5.png">
-----------------------------------------
**To test that you are still able to create and update Broadcasts, you can do the following:**
- Navigate to `/internal/broadcasts`.
- Create a `Broadcast` of either `broadcastable_type` (`Announcement` or `WelcomeNotification`), or of both to ensure that you can create _any_ type of Broadcast:
<img width="1173" alt="Screen Shot 2020-07-27 at 9 52 12 AM" src="https://user-images.githubusercontent.com/32834804/88563173-dfac5280-cfee-11ea-8578-66ae3e240e32.png">

- Update one of your existing `Broadcast`s to ensure that the update method works properly with the polymorphic changes:

<img width="1173" alt="Screen Shot 2020-07-27 at 9 54 48 AM" src="https://user-images.githubusercontent.com/32834804/88563454-3ade4500-cfef-11ea-86e6-61491ce92ead.png">

- Try to break things! 🛠️ 

## Added tests?

- [x] yes: added additional specs and refactored existing specs

## Added to documentation?


## [optional] Are there any post-deployment tasks we need to perform?
There is a `data_update` script included in this PR to port over the data for `broadcastable` and `banner_style`.

## [optional] What gif best describes this PR or how it makes you feel?

![The Office](https://media.giphy.com/media/13jghlUIB6FHZm/giphy.gif)
